service/controller/resource/instancestatus/resource.go

```diff
@@ -0,0 +1,143 @@
+package instancestatus
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/giantswarm/apiextensions/pkg/clientset/versioned"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+
+	"github.com/giantswarm/aws-operator/service/controller/controllercontext"
+	"github.com/giantswarm/aws-operator/service/controller/key"
+)
+
+const (
+	Name = "instancestatus"
+)
+
+type Config struct {
+	G8sClient versioned.Interface
+	Logger    micrologger.Logger
+}
+
+type Resource struct {
+	g8sClient versioned.Interface
+	logger    micrologger.Logger
+}
+
+func New(config Config) (*Resource, error) {
+	if config.G8sClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.G8sClient must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	r := &Resource{
+		g8sClient: config.G8sClient,
+		logger:    config.Logger,
+	}
+
+	return r, nil
+}
+
+func (r *Resource) Name() string {
+	return Name
+}
+
+func (r *Resource) ensure(ctx context.Context, obj interface{}) error {
+	cr, err := key.ToMachineDeployment(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+	cc, err := controllercontext.FromContext(ctx)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	i := &ec2.DescribeInstancesInput{
+		Filters: []*ec2.Filter{
+			{
+				Name: aws.String(fmt.Sprintf("tag:%s", key.TagCluster)),
+				Values: []*string{
+					aws.String(key.ClusterID(&cr)),
+				},
+			},
+			{
+				Name: aws.String(fmt.Sprintf("tag:%s", key.TagMachineDeployment)),
+				Values: []*string{
+					aws.String(key.MachineDeploymentID(&cr)),
+				},
+			},
+			{
+				Name: aws.String(fmt.Sprintf("tag:%s", key.TagStack)),
+				Values: []*string{
+					aws.String(key.StackTCNP),
+				},
+			},
+			{
+				Name: aws.String("instance-state-name"),
+				Values: []*string{
+					aws.String(ec2.InstanceStateNamePending),
+					aws.String(ec2.InstanceStateNameRunning),
+					aws.String(ec2.InstanceStateNameStopped),
+					aws.String(ec2.InstanceStateNameStopping),
+				},
+			},
+		},
+	}
+
+	o, err := cc.Client.TenantCluster.AWS.EC2.DescribeInstances(i)
+	if IsNotFound(err) {
+		r.logger.LogCtx(ctx, "level", "debug", "message", "worker asg not available yet")
```

Comment:
I think we already have something that more or less does what we need here. See https://github.com/giantswarm/aws-operator/blob/master/service/controller/resource/asgstatus/resource.go. We should maybe rather extend the existing resource than making up a new one.

Reply:
We decided to create a new resource because the `asg` resource is strictly looking for information related to the `asg`.

The new resource is only concerned with the `instances` inside the `asg`. It's true that they work pretty similarly but mixing it in one resources created a mess. 

We can maybe think about some better abstraction here - but keeping it simple made sense to me.

Reply:
I am worried we do this another 3 times and get more rate limit issues again. I would like to prevent this.

Reply:
I will add a TODO for myself to refactor this into one resource fetching from the AWS API and then the others reading this from the controllercontext.